patches/server/1039-Add-LivingEntity-getPotionEffects.patch

```diff
@@ -0,0 +1,53 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Tamion <70228790+notTamion@users.noreply.github.com>
+Date: Thu, 12 Oct 2023 18:34:08 +0200
+Subject: [PATCH] Add LivingEntity#getPotionEffects
+
+
+diff --git a/src/main/java/net/minecraft/world/effect/MobEffectInstance.java b/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
+index 68e1b8271475996020af50b3b2cf04cd25aa6c85..528541288f0656b42c8a60e0555a8959f24a7b8b 100644
+--- a/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
++++ b/src/main/java/net/minecraft/world/effect/MobEffectInstance.java
+@@ -327,6 +327,17 @@ public class MobEffectInstance implements Comparable<MobEffectInstance> {
+         return (this.getDuration() <= 32147 || mobEffectInstance.getDuration() <= 32147) && (!this.isAmbient() || !mobEffectInstance.isAmbient()) ? ComparisonChain.start().compareFalseFirst(this.isAmbient(), mobEffectInstance.isAmbient()).compareFalseFirst(this.isInfiniteDuration(), mobEffectInstance.isInfiniteDuration()).compare(this.getDuration(), mobEffectInstance.getDuration()).compare(this.getEffect().getColor(), mobEffectInstance.getEffect().getColor()).result() : ComparisonChain.start().compare(this.isAmbient(), mobEffectInstance.isAmbient()).compare(this.getEffect().getColor(), mobEffectInstance.getEffect().getColor()).result();
+     }
+ 
++    // Paper start
++    @org.jetbrains.annotations.NotNull
++    public java.util.List<MobEffectInstance> getHiddenEffects() {
++        java.util.List<MobEffectInstance> effects = new java.util.ArrayList<>();
++        for (MobEffectInstance effect = this.hiddenEffect; effect != null; effect = effect.hiddenEffect) {
++            effects.add(effect);
++        }
++        return effects;
++    }
++    // Paper end
++
+     public static class FactorData {
+         public static final Codec<MobEffectInstance.FactorData> CODEC = RecordCodecBuilder.create((instance) -> {
+             return instance.group(ExtraCodecs.NON_NEGATIVE_INT.fieldOf("padding_duration").forGetter((data) -> {
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+index daadfec2d86f7957072a639e1e36d4082448f35f..d7d6837129d555cbaf5341fcb1f39610fc1fa51f 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+@@ -503,6 +503,20 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
+         return effects;
+     }
+ 
++    // Paper start
++    @Override
++    public Collection<PotionEffect> getAllPotionEffects() {
++        List<PotionEffect> effects = new ArrayList<PotionEffect>();
```

Comment:
The generic type <PotionEffect> is not needed in the right assignment, also in this part you skip completely the showIcon part that could be relevant on player with custom made potion.

Reply:
what exactly do you mean by the showIcon part? i assume that variable does exactly what its name suggests and i don't see how exactly that would be relevant for this API method

Reply:
It's a nbt key/field, when used to create a potion using a command or the api, the icon in the top right corner can be hidden in game (not in the inventory). By default it's set to follow the show_particles parameter but it could be different for custom effect. However in your method you ignore that, and so for those type of potions the getter (PotionEffect#hasIcon) will be wrong. It's really niche api but there's no reason to ignore it i think.

Reply:
ah thanks. i used the getActivePotionEffects method as a starting point so i used the wrong overload